### ISSUE TITLE
Using `set_all_bn_inputs_no_bias` with a `repeat` layer as input.

### ISSUE THREAD

**Issue by arrufat** [davisking/dlib#2216]
I was trying to define a naive version of DenseNets. In those networks after each dense block there's a transition module that has a batchnormalization layer as its first step.

When using `repeat` layers as an input of a `bn_` layer, and calling `set_all_bn_inputs_no_bias` gives me the following error:

```
../../external/dlib/dlib/../dlib/dnn/layers.h:1593:41: error: ‘dlib::add_layer<dlib::bn_<dlib::CONV_MODE>, dlib::repeat<1, con1x1, dlib::add_layer<dlib::affine_, dlib::input_rgb_image, void> >, void>::subnet_type’ {aka ‘class dlib::repeat<1, con1x1, dlib::add_layer<dlib::affine_, dlib::input_rgb_image, void> >’} has no member named ‘layer_details’; did you mean ‘layer_details_type’?
 1593 |                 disable_bias(l.subnet().layer_details());
      |                              ~~~~~~~~~~~^~~~~~~~~~~~~
      |                              layer_details_type
../../external/dlib/dlib/../dlib/dnn/layers.h:1594:62: error: ‘dlib::add_layer<dlib::bn_<dlib::CONV_MODE>, dlib::repeat<1, con1x1, dlib::add_layer<dlib::affine_, dlib::input_rgb_image, void> >, void>::subnet_type’ {aka ‘class dlib::repeat<1, con1x1, dlib::add_layer<dlib::affine_, dlib::input_rgb_image, void> >’} has no member named ‘layer_details’; did you mean ‘layer_details_type’?
 1594 |                 set_bias_learning_rate_multiplier(l.subnet().layer_details(), 0);
      |                                                   ~~~~~~~~~~~^~~~~~~~~~~~~
      |                                                   layer_details_type
../../external/dlib/dlib/../dlib/dnn/layers.h:1595:61: error: ‘dlib::add_layer<dlib::bn_<dlib::CONV_MODE>, dlib::repeat<1, con1x1, dlib::add_layer<dlib::affine_, dlib::input_rgb_image, void> >, void>::subnet_type’ {aka ‘class dlib::repeat<1, con1x1, dlib::add_layer<dlib::affine_, dlib::input_rgb_image, void> >’} has no member named ‘layer_details’; did you mean ‘layer_details_type’?
 1595 |                 set_bias_weight_decay_multiplier(l.subnet().layer_details(), 0);
      |                                                  ~~~~~~~~~~~^~~~~~~~~~~~~
      |                                                  layer_details_type
ninja: build stopped: subcommand failed.
```

I've made a simple example to illustrate the issue (note that I needed to use the `affine` layer to be able to call `repeat`, as it doesn't accept `input` layers, this trick also solves my issue):

``` c++
using namespace dlib;
template <typename SUBNET> using con1x1 = con<1, 1, 1, 1, 1, SUBNET>;

int main()
{
    using net_type = bn_con<repeat<1, con1x1, affine<input_rgb_image>>>;
    set_all_bn_inputs_no_bias(net);
}
```

I was wondering if something could be done about it. Thanks in advance :)

**Comment by davisking** [comment]
Eh, yeah, `repeat` is special.  I kinda forgot about `repeat`. It's an array of layer detail objects basically.  So it requires special handling.

**Comment by davisking** [comment]
Like you just need another overload for `repeat` layers that calls `get_repeated_layer()` instead of `layer_details()`.

**Comment by arrufat** [comment]
Thank you, I've added this overload:

``` c++
template <layer_mode mode, size_t N, template <typename> class R, typename U, typename E>
void set_input_no_bias(add_layer<bn_<mode>, repeat<N, R, U>, E>& l)
{
    disable_bias(l.subnet().get_repeated_layer(0).layer_details());
    set_bias_learning_rate_multiplier(l.subnet().get_repeated_layer(0).layer_details(), 0);
    set_bias_weight_decay_multiplier(l.subnet().get_repeated_layer(0).layer_details(), 0);
}
```

And it works for the case that I show above. But not when I repeat several dense_layers:

``` c++
using namespace dlib;
template <long num_filters, long ks, int s, typename SUBNET>
using con_block = relu<bn_con<add_layer<con_<num_filters, ks, ks, s, s, ks/2, ks/2>, SUBNET>>>;

template <typename INPUT>
using stem = add_layer<max_pool_<3, 3, 2, 2, 1, 1>, con_block<64, 7, 2, INPUT>>;

template <long num_filters, long growth_rate, typename SUBNET>
using dense_layer = concat2<tag1, tag2,
                    tag2<con_block<growth_rate, 3, 1,
                    con_block<num_filters, 1, 1,
                    tag1<relu<bn_con<SUBNET>>>>>>>;

template <typename SUBNET>
using dense_layer_128 = dense_layer<128, 32, SUBNET>;

int main()
{
    using net_type = repeat<2, dense_layer_128, stem<input_rgb_image>>;
    set_all_bn_inputs_no_bias(net);
}
```

I still get the same error. Am I missing another overload?

**Comment by davisking** [comment]
Not sure, what's the error you get?  On the surface what you have looks right.

**Comment by arrufat** [comment]
I am getting the same error as before, it seems that in the case above, that overload is not being used...

```
/home/adria/Projects/dlib/dlib/../dlib/dnn/layers.h:1736:41: error: ‘dlib::add_layer<dlib::bn_<dlib::CONV_MODE>, dlib::impl::repeat_input_layer, void>::subnet_type’ {aka ‘class dlib::impl::repeat_input_layer’} has no member named ‘layer_details’
 1736 |                 disable_bias(l.subnet().layer_details());
      |                              ~~~~~~~~~~~^~~~~~~~~~~~~
/home/adria/Projects/dlib/dlib/../dlib/dnn/layers.h:1737:62: error: ‘dlib::add_layer<dlib::bn_<dlib::CONV_MODE>, dlib::impl::repeat_input_layer, void>::subnet_type’ {aka ‘class dlib::impl::repeat_input_layer’} has no member named ‘layer_details’
 1737 |                 set_bias_learning_rate_multiplier(l.subnet().layer_details(), 0);
      |                                                   ~~~~~~~~~~~^~~~~~~~~~~~~
/home/adria/Projects/dlib/dlib/../dlib/dnn/layers.h:1738:61: error: ‘dlib::add_layer<dlib::bn_<dlib::CONV_MODE>, dlib::impl::repeat_input_layer, void>::subnet_type’ {aka ‘class dlib::impl::repeat_input_layer’} has no member named ‘layer_details’
 1738 |                 set_bias_weight_decay_multiplier(l.subnet().layer_details(), 0);
      |                                                  ~~~~~~~~~~~^~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
```

**Comment by arrufat** [comment]
Ah, it's because I have a `tag` layer after the `bn_con`. I will try to fix it.

**Comment by davisking** [comment]
Yeah, I was just about to say that repeat blocks that start with tag layers are treated specially.  That's what that `repeat_input_layer` is about.  It converts a tag layer into something that avoids the implied tensor copy that would otherwise be there.  But you have to handle that funny layer type when it gets visited.

**Comment by arrufat** [comment]
Yes, I am trying to play with the overloads with no luck, yet...
Since the error complains about this type:
```
dlib::add_layer<dlib::bn_<dlib::CONV_MODE>, dlib::add_tag_layer<1, dlib::impl::repeat_input_layer, void>, void>
```
I added this, but not working yet.
``` c++
template <layer_mode mode, unsigned long ID, size_t N, template <typename> class R, typename U, typename E>
void disable_input_bias(add_layer<bn_<mode>, add_tag_layer<ID, impl::repeat_input_layer, U>, E>& l)
{
    disable_bias(l.subnet().subnet().get_repeated_layer(0).layer_details());
    set_bias_learning_rate_multiplier(l.subnet().subnet().get_repeated_layer(0).layer_details(), 0);
    set_bias_weight_decay_multiplier(l.subnet().subnet().get_repeated_layer(0).layer_details(), 0); 
}
```

**Comment by arrufat** [comment]
Done, I forgot to delete the unused template parameters 😅

**Comment by arrufat** [comment]
Closed with #2213. Also, `set_all_bn_inputs_no_bias` has been renamed to `disable_duplicative_bias`.
